proposals/0000-char-kind.rst

```diff
@@ -0,0 +1,315 @@
+.. ticket-url:: `#11342 <https://gitlab.haskell.org/ghc/ghc/-/issues/11342>`_
+.. implemented:: `!4351 <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4351>`_
+.. highlight:: haskell
+.. contents::
+
+Haskell has support for type-level data in the form of the ``DataKinds``
+extension, which promotes algebraic data types such as ``Bool``, ``Maybe``, or
+even user-defined ADTs. It also has special support for two non-algebraic
+data types: ``Symbol`` and ``Natural``, representing type-level strings and
+type-level numbers respectively.
+
+We propose to add support for one more non-algebraic data type, namely
+``Char``, thereby resolving the long-standing issue `#11342
+<https://gitlab.haskell.org/ghc/ghc/-/issues/11342>`_.
+
+Motivation
+----------
+
+The existing support for type-level strings is limited to:
+
+1. String literals::
+
+     ghci> :kind "myString"
+     "myString" :: Symbol
+
+2. String concatenation::
+
+     ghci> :kind! AppendSymbol "hello" "world"
+     AppendSymbol "hello" "world" :: Symbol
+     = "helloworld"
+
+3. String comparison::
+
+     ghci> :kind! CmpSymbol "hello" "world"
+     CmpSymbol "hello" "world" :: Ordering
+     = 'LT
+
+There are no ways to decompose or analyse a type-level string in terms of its
+constituent characters. Constrast that with the API of `Text
+<https://hackage.haskell.org/package/text/docs/Data-Text.html>`_, which offers
+a multitude of functions such as ``uncons``, ``map``, ``splitAt``, and so on.
+
+We could try to extend the API of ``Symbol`` accordingly. For example, the
+type-level counterpart of ``uncons :: Text -> Maybe (Char, Text)`` could be a
+built-in type family such as ``UnconsSymbol :: Symbol -> Maybe (Char,
+Symbol)``.
+
+Notice that the return type of the proposed ``UnconsSymbol`` mentions ``Char``.
+However, there's currently no support for type-level characters::
+
+  ghci> :kind! 'x'
+  <interactive>:1:1: error: parse error on input ‘'’
+
+We propose to fix this omission. The ``Char`` kind and the accompanying
+built-in type families will make it possible to implement type-level parsers
+(see the "Examples" section below).
+
+Proposed Change Specification
+-----------------------------
+
+1. Extend the grammar of type-level literals
+   with character literals::
+
+     tylit ::=
+         INTEGER
+       | STRING
+       | CHAR       (NEW)
+
+   The lexical syntax matches that of term-level character literals: the
+   character enclosed in single quotes, e.g. ``'X'`` or ``'\n'``.
+
+2. Extend the ``GHC.TypeLits`` module
+   with the following built-in type families::
+
+     type family CmpChar (a :: Char) (b :: Char) :: Ordering
+     type family ConsSymbol (a :: Char) (b :: Symbol) :: Symbol
+     type family UnconsSymbol (a :: Symbol) :: Maybe (Char, Symbol)
+
+   * The semantics of ``CmpChar`` match that of ``compare @Char``.
+   * The semantics of ``ConsSymbol`` and ``UnconsSymbol`` match that of
+     ``Data.Text.cons`` and ``Data.Text.uncons`` respectively.
```

Comment:
What about surrogates?

```
Prelude Data.Text> cons '\xD800' "foo"
"\65533foo"
Prelude Data.Text> 0xd800
55296
```

Reply:
Very interesting. I didn't know about this, but here's a relevant excerpt from `text` sources:

```
-- | Map a 'Char' to a 'Text'-safe value.
--
-- UTF-16 surrogate code points are not included in the set of Unicode
-- scalar values, but are unfortunately admitted as valid 'Char'
-- values by Haskell.  They cannot be represented in a 'Text'.  This
-- function remaps those code points to the Unicode replacement
-- character (U+FFFD, \'&#xfffd;\'), and leaves other code points
-- unchanged.
safe :: Char -> Char
safe c
    | ord c .&. 0x1ff800 /= 0xd800 = c
    | otherwise                    = '\xfffd'
```

I think `ConsSymbol` should not reduce (i.e. it should get stuck) in this case. Either that, or a produce a proper type error.

Reply:
FWIW, `Symbol` is not `Text`, so it can contain surrogate codepoints, where (safely) constructed `Text` cannot. `Symbol` is really `String` at the moment i.e a list of (possibly invalid) codepoints/`Char`s.

```haskell
Prelude GHC.TypeLits Data.Text> "\xD800" :: String
"\55296"
Prelude GHC.TypeLits Data.Text> "\xD800" :: Text
"\65533"
Prelude GHC.TypeLits Data.Text> case someSymbolVal "\xD800" of SomeSymbol p -> symbolVal p
"\55296"
```

Reply:
... so IMHO just going with ` (:)` and `uncons` semantics is fine IMHO. But if someone wants to fix `Symbol` to represent only correct Unicode text, that would be another proposal...

Reply:
> `Symbol` is really `String` at the moment

Internally, `Symbol` is represented as a `FastString`, defined thus:

```
{-| A 'FastString' is a UTF-8 encoded string together with a unique ID. All
'FastString's are stored in a global hashtable to support fast O(1)
comparison.

It is also associated with a lazy reference to the Z-encoding
of this string which is used by the compiler internally.
-}
data FastString = FastString {
      uniq    :: {-# UNPACK #-} !Int, -- unique id
      n_chars :: {-# UNPACK #-} !Int, -- number of chars
      fs_sbs  :: {-# UNPACK #-} !ShortByteString,
      fs_zenc :: FastZString
      -- ^ Lazily computed Z-encoding of this string. See Note [Z-Encoding] in
      -- GHC.Utils.Encoding.
      --
      -- Since 'FastString's are globally memoized this is computed at most
      -- once for any given string.
  }
```

Note that it says "UTF-8 encoded". So either the implementation of `Symbol` has a bug (and it shouldn't use `FastString`), or its surface API has a bug (and it shouldn't represent byte sequences which are not valid UTF-8).

Reply:
*UTF-8 encoded* is not correct. It's a Modified UTF-8 (= it doesn't contain zero-character). And it also can contain surrogates.

Modified-UTF8:

```haskell
-- | Gives the Modified UTF-8 encoded bytes corresponding to a 'FastString'
bytesFS, fastStringToByteString :: FastString -> ByteString
bytesFS = fastStringToByteString

{-# DEPRECATED fastStringToByteString "Use `bytesFS` instead" #-}
fastStringToByteString f = SBS.fromShort $ fs_sbs f

fastStringToShortByteString :: FastString -> ShortByteString
fastStringToShortByteString = fs_sbs
```

Surrogates: Because it has to, because non-abstract `String = [Char]` can contain them.

Reply:
https://www.unicode.org/versions/Unicode13.0.0/ch03.pdf

Page 77.

> C1) A process shall not interpret a high-surrogate code point or a low-surrogate code point as an abstract character.
> 
> • The high-surrogate and low-surrogate code points are designated for surrogate code units in the UTF-16 character encoding form. They are unassigned to any abstract character.

I'm not sure if UTF-16 can even represent surrogate codepoints (which `text` uses atm).

---

This problem is very  deep in Haskell / GHC.

String literals like "\xD800" are wrong to begin with, they are not *Unicode Text*. Yet Haskell Report doesn't claim they are, so we shouldn't here either.

We can have debate elsewhere whether this trade-off  of "string literals are not unicode text because we need to represent also binary yadda-yadda" elsewhere. More productive (short-term) is just accept that strings in "core" Haskell are not "Unicode Text", but a list of codepoints (including non-abstract characters).

I speak from the experience of low-level problems these kind of hacks impose. For example, if `text` were to change to UTF-8 represention internally tomorrow, it would still need to do conversion from FastString Modified-UTF8 to recode zero-character and filter surrogates, so literals won't become any cheaper.

Reply:
OK. So `Symbol` should be isomorphic to `[Char]`, and the semantics of `ConsSymbol` and `UnconsSymbol` should match that of `(:)` and [`GHC.List.uncons`](https://hackage.haskell.org/package/base-4.14.1.0/docs/GHC-List.html#v:uncons). At the same time, internally it's a `FastString` that stores Modified UTF-8, therefore its performance characteristics are different from that of `[Char]`.

Is that right?

Reply:
@int-index Yes, that is right.

Reply:
Has it been ever proposed that we fix `Char`? `ByteString` `[WordXx]` etc. are there to handle the other cases, IMO.

Reply:
On the other hand, https://simonsapin.github.io/wtf-8/ is how rust encodes surrogate codepoints as an extension of UTF-8. This could be used for fast string.

Reply:
Hmmmm
```
ghci> :set -XDataKinds
ghci> :set -XTypeFamilies
ghci> import GHC.TypeLits
ghci> type X = ConsSymbol '\xD800' "foo"
ghci> :kind! X
X :: Symbol
= "\55296foo"
```

Reply:
If I'm not mistaken, there exist strings with no counterparts in `FastString`. Is it crucial? If so, it seems like a problem for a new issue.

Reply:
> Has it been ever proposed that we fix Char?
How exactly?

Reply:
Now we have the following test case
```
testUncons :: ConsSymbol '\xD800' "foo" :~: "\55296foo"
testUncons = Refl
```

Reply:
> > Has it been ever proposed that we fix Char?
> 
> How exactly?

By only allowing valid unicode.

Reply:
@Ericson2314 it's a swamp, e.g. `FilePath = String` forces us to have 'leaky" strings, which allow invalid unicode - because filepaths could be such.

https://github.com/rust-lang/rust/issues/12056 is "fun" read of some issues to consider...

Reply:
Well, time to advocate for `FilePath != String` then :D

Reply:
> `FilePath = String` forces us to have 'leaky" strings

I thought filepath `ByteString -> String` decoding was using some weird rounde-trippable encoding that essentially allows to represent any sequence of bytes as a sequence of Unicode codepoints? I.e. the fact that a `FilePath` should have been a `ByteString` is an entirely different (and already solved) issue.